Description: Keep dhclient as default client
Author: Brett Holman <brett.holman@canonical.com>
Last-Update: 2024-02-20

--- a/config/cloud.cfg.tmpl
+++ b/config/cloud.cfg.tmpl
@@ -320,7 +320,7 @@ system_info:
 {% elif variant in ["ubuntu", "unknown"] %}
 {# SRU_BLOCKER: do not ship network renderers on Xenial, Bionic or Eoan #}
   network:
-    dhcp_client_priority: [dhcpcd, dhclient, udhcpc]
+    dhcp_client_priority: [dhclient, dhcpcd, udhcpc]
     renderers: ['netplan', 'eni', 'sysconfig']
     activators: ['netplan', 'eni', 'network-manager', 'networkd']
 {% elif is_rhel %}
