Description: Handle symlinks as devices in mount_cb
Author: Daniel Watkins <daniel.watkins@canonical.com>
Origin: upstream, http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/revision/1139
Bug: https://bugs.launchpad.net/cloud-init/+bug/1490796
--- a/cloudinit/util.py
+++ b/cloudinit/util.py
@@ -804,8 +804,8 @@
         mounted[dev] = (dev, fstype, mp, False)
 
     umount = False
-    if device in mounted:
-        mountpoint = "%s/" % mounted[device][2]
+    if os.path.realpath(device) in mounted:
+        mountpoint = "%s/" % mounted[os.path.realpath(device)][2]
     else:
         tmpd = tempfile.mkdtemp()
 
