Description: Enable backports apt pocket
Author: Ben Howard <ben.howard@canonical.com>
Bug: https://bugs.launchpad.net/cloud-init/+bug/1177432
--- a/templates/sources.list.ubuntu.tmpl
+++ b/templates/sources.list.ubuntu.tmpl
@@ -35,15 +35,22 @@
 # deb $mirror $codename-updates multiverse
 # deb-src $mirror $codename-updates multiverse
 
-\## Uncomment the following two lines to add software from the 'backports'
-\## repository.
 \## N.B. software from this repository may not have been tested as
 \## extensively as that contained in the main release, although it includes
 \## newer versions of some applications which may provide useful features.
 \## Also, please note that software in backports WILL NOT receive any review
 \## or updates from the Ubuntu security team.
-# deb $mirror $codename-backports main restricted universe multiverse
-# deb-src $mirror $codename-backports main restricted universe multiverse
+deb $mirror $codename-backports main
+deb-src $mirror $codename-backports main
+deb $mirror $codename-backports universe
+deb-src $mirror $codename-backports universe
+
+\## Uncomment the following lines to add software from the 'backports'
+\## repository for mulitverse and restricted.
+# deb $mirror $codename-backports multiverse
+# deb-src $mirror $codename-backports multiverse
+# deb $mirror $codename-backports restricted
+# deb-src $mirror $codename-backports restricted
 
 \## Uncomment the following two lines to add software from Canonical's
 \## 'partner' repository.
